security/src/main/java/io/airlift/security/client/SpnegoAuthentication.java

```diff
@@ -0,0 +1,92 @@
+package io.airlift.security.client;
+
+import io.airlift.log.Logger;
+import io.airlift.security.exception.AuthenticationException;
+import io.airlift.security.utils.CloseableGSSContext;
+import org.eclipse.jetty.client.api.Authentication;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.util.Attributes;
+import org.eclipse.jetty.util.B64Code;
+import org.ietf.jgss.GSSException;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import static java.util.Objects.requireNonNull;
+
+public class SpnegoAuthentication
+        implements Authentication
+{
+    private static final String NEGOTIATE = HttpHeader.NEGOTIATE.asString();
+    private static final Logger log = Logger.get(SpnegoAuthentication.class);
+    private final String serviceName;
+
+    public SpnegoAuthentication(String krb5ConfPath, String serviceName)
+    {
+        requireNonNull(krb5ConfPath, "krb5ConfPath is null");
+        this.serviceName = requireNonNull(serviceName, "serviceName is null");
+
+        System.setProperty("java.security.krb5.conf", krb5ConfPath);
+    }
+
+    @Override
+    public Result authenticate(Request request, ContentResponse response, HeaderInfo headerInfo, Attributes context)
+    {
+        URI requestURI = request.getURI();
+        try (CloseableGSSContext gssContext = new CloseableGSSContext(serviceName, requestURI)) {
+            // We generate the token once and send it to the peer.
+            byte[] token = gssContext.getContext().initSecContext(new byte[0], 0, 0);
+            if (token != null) {
+                log.debug("Successfully established GSSContext with source name: %s and target name: %s",
+                        gssContext.getContext().getSrcName(),
+                        gssContext.getContext().getTargName());
+                String tokenValue = String.format("%s %s", NEGOTIATE,
+                        String.valueOf(B64Code.encode(token)));
+                URI baseUri = new URI(requestURI.getScheme(), null, requestURI.getHost(), requestURI.getPort(),
+                        null, null, null);
```

Comment:
Isn't this the same logic as `normalizedUri`? Is there any relationship between these URIs and those handled by the authentication store? If so, pull this out into a separate method that can be used from both places.